File: clients/ios/Lextures/Core/LMS/ModuleLastVisited.swift
import Foundation

struct LastVisitedModuleEntry: Codable, Equatable {
    var itemId: String
    var kind: String
    var title: String
    var openedAt: String
}

/// Per-course last opened module item (parity with web `last-visited-module-item.ts`).
enum ModuleLastVisited {
    private static let storageKey = "lextures:last-module-item:v1"

    static func record(courseCode: String, itemId: String, kind: String, title: String) {
        let code = courseCode.trimmingCharacters(in: .whitespacesAndNewlines)
        let id = itemId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !code.isEmpty, !id.isEmpty else { return }

        var store = readStore()
        store[code] = LastVisitedModuleEntry(
            itemId: id,
            kind: kind,
            title: title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "Untitled" : title,
            openedAt: ISO8601DateFormatter().string(from: Date())
        )
        writeStore(store)
    }

    static func entry(for courseCode: String) -> LastVisitedModuleEntry? {
        let code = courseCode.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !code.isEmpty else { return nil }
        return readStore()[code]
    }

    private static func readStore() -> [String: LastVisitedModuleEntry] {
        guard let data = UserDefaults.standard.data(forKey: storageKey) else { return [:] }
        return (try? JSONDecoder().decode([String: LastVisitedModuleEntry].self, from: data)) ?? [:]
    }

    private static func writeStore(_ store: [String: LastVisitedModuleEntry]) {
        guard let data = try? JSONEncoder().encode(store) else { return }
        UserDefaults.standard.set(data, forKey: storageKey)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/ModuleLastVisited.kt
package com.lextures.android.core.lms

import android.content.Context
import java.time.Instant
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json

@Serializable
data class LastVisitedModuleEntry(
    val itemId: String,
    val kind: String,
    val title: String,
    val openedAt: String,
)

/** Per-course last opened module item (parity with web `last-visited-module-item.ts`). */
object ModuleLastVisited {
    private const val STORAGE_KEY = "lextures:last-module-item:v1"
    private val json = Json { ignoreUnknownKeys = true }
    private val storeSerializer = MapSerializer(String.serializer(), LastVisitedModuleEntry.serializer())

    fun record(context: Context, courseCode: String, itemId: String, kind: String, title: String) {
        val code = courseCode.trim()
        val id = itemId.trim()
        if (code.isEmpty() || id.isEmpty()) return

        val store = readStore(context).toMutableMap()
        store[code] = LastVisitedModuleEntry(
            itemId = id,
            kind = kind,
            title = title.trim().ifEmpty { "Untitled" },
            openedAt = Instant.now().toString(),
        )
        writeStore(context, store)
    }

    fun entry(context: Context, courseCode: String): LastVisitedModuleEntry? {
        val code = courseCode.trim()
        if (code.isEmpty()) return null
        return readStore(context)[code]
    }

    private fun readStore(context: Context): Map<String, LastVisitedModuleEntry> {
        val raw = context.getSharedPreferences("module_last_visited", Context.MODE_PRIVATE)
            .getString(STORAGE_KEY, null)
            ?: return emptyMap()
        return runCatching {
            json.decodeFromString(storeSerializer, raw)
        }.getOrDefault(emptyMap())
    }

    private fun writeStore(context: Context, store: Map<String, LastVisitedModuleEntry>) {
        val encoded = json.encodeToString(storeSerializer, store)
        context.getSharedPreferences("module_last_visited", Context.MODE_PRIVATE)
            .edit()
            .putString(STORAGE_KEY, encoded)
            .apply()
    }
}